File: dashboard/server/routes/apps.ts
import { Router } from "express";
import db from "../db.js";

const router = Router();

router.get("/", (req, res) => {
  const from = (req.query.from as string) || new Date().toISOString().slice(0, 10);
  const to = (req.query.to as string) || new Date().toISOString().slice(0, 10);
  const limit = parseInt((req.query.limit as string) || "20", 10);

  const fromDate = `${from}T00:00:00`;
  const toDate = `${to}T23:59:59`;

  const rows = db
    .prepare(
      `SELECT
         COALESCE(e.app, 'Unknown') as app,
         COALESCE(c.name, 'uncategorized') as category,
         SUM(e.duration) as total_seconds,
         COUNT(*) as event_count
       FROM events e
       LEFT JOIN categories c ON e.category_id = c.id
       WHERE e.timestamp >= ? AND e.timestamp <= ? AND e.is_afk = 0
       GROUP BY e.app
       ORDER BY total_seconds DESC
       LIMIT ?`
    )
    .all(fromDate, toDate, limit) as Array<{
    app: string;
    category: string;
    total_seconds: number;
    event_count: number;
  }>;

  const totalSeconds = rows.reduce((sum, r) => sum + r.total_seconds, 0);

  const apps = rows.map((r) => ({
    app: r.app,
    category: r.category,
    seconds: Math.round(r.total_seconds),
    time: formatDuration(r.total_seconds),
    pct: totalSeconds > 0 ? Math.round((r.total_seconds / totalSeconds) * 1000) / 10 : 0,
    events: r.event_count,
  }));

  res.json(apps);
});

function formatDuration(seconds: number): string {
  const h = Math.floor(seconds / 3600);
  const m = Math.round((seconds % 3600) / 60);
  if (h > 0) return `${h}h ${m}m`;
  return `${m}m`;
}

export default router;

File: src/query/apps.rs
use rusqlite::Connection;
use crate::error::Result;
use crate::types::{format_duration, AppBreakdown};

pub fn build_apps(
    conn: &Connection,
    from_date: &str,
    to_date: &str,
    limit: i64,
) -> Result<Vec<AppBreakdown>> {
    let mut stmt = conn.prepare(
        "SELECT
           CASE
             WHEN e.url_domain IS NOT NULL AND e.url_domain != ''
             THEN e.url_domain
             ELSE COALESCE(e.app, 'Unknown')
           END as label,
           COALESCE(c.name, 'uncategorized') as category,
           SUM(e.duration) as total_seconds,
           COUNT(*) as event_count
         FROM events e
         LEFT JOIN categories c ON e.category_id = c.id
         WHERE e.timestamp >= ?1 AND e.timestamp <= ?2 AND e.is_afk = 0 AND e.duration > 0
         GROUP BY label
         ORDER BY total_seconds DESC
         LIMIT ?3",
    )?;

    let rows = stmt.query_map(rusqlite::params![from_date, to_date, limit], |row| {
        Ok((
            row.get::<_, String>(0)?,
            row.get::<_, String>(1)?,
            row.get::<_, f64>(2)?,
            row.get::<_, i64>(3)?,
        ))
    })?;

    let mut data = Vec::new();
    let mut total_seconds: f64 = 0.0;
    for row in rows {
        let r = row?;
        total_seconds += r.2;
        data.push(r);
    }

    let apps = data
        .into_iter()
        .map(|(app, category, secs, events)| {
            let pct = if total_seconds > 0.0 {
                (secs / total_seconds * 1000.0).round() / 10.0
            } else {
                0.0
            };
            AppBreakdown {
                app,
                category,
                seconds: secs.round() as i64,
                time: format_duration(secs),
                pct,
                events,
            }
        })
        .collect();

    Ok(apps)
}
